File: frontend/src/services/files.ts
import { api } from './api'

export interface FileRecord {
  id: string
  project_id: string
  user_id: string
  project_name?: string | null
  file_name: string
  file_type: string
  storage_key: string
  parse_status: string
  source_type?: string | null
  file_size?: number | null
  metadata_json?: string | null
}

const NON_INPUT_SOURCE_TYPES = new Set(['step1_export_final', 'step1_draft_commit'])

export function isInputProjectFile(file: FileRecord) {
  const sourceType = (file.source_type || '').trim()
  if (sourceType && NON_INPUT_SOURCE_TYPES.has(sourceType)) return false
  if (file.file_type === 'draft-json') return false
  return true
}

export interface FileCreateRequest {
  user_id?: string
  project_name?: string | null
  file_name: string
  file_type: string
  storage_key?: string
  source_type?: string | null
  file_size?: number | null
  metadata_json?: string | null
  draft_thread_id?: string | null
  draft_payload?: Record<string, unknown> | null
}

export async function listFiles(projectId: string) {
  const { data } = await api.get<{ items: FileRecord[] }>(`/files/${projectId}`)
  return data.items
}

export async function createFileRecord(projectId: string, payload: FileCreateRequest) {
  const { data } = await api.post<FileRecord>(`/files/${projectId}`, payload)
  return data
}

export async function uploadProjectFile(projectId: string, file: File, userId = 'demo-user-id') {
  const form = new FormData()
  form.append('upload', file)
  const { data } = await api.post<FileRecord>(`/files/${projectId}/upload`, form, {
    params: { user_id: userId },
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function deleteFile(projectId: string, fileId: string) {
  const { data } = await api.delete(`/files/${projectId}/${fileId}`)
  return data
}

File: Backend/app/api/v1/routes/files.py
from pathlib import Path
from uuid import uuid4

from typing import Any

from fastapi import APIRouter, Body, Depends, File, UploadFile
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.core.errors import not_found
from app.db.session import get_db
from app.services.file_service import FileService

router = APIRouter()


class FileCreateRequest(BaseModel):
    user_id: str = Field(default="demo-user-id")
    project_name: str | None = None
    file_name: str = Field(default="unnamed", min_length=1)
    file_type: str = Field(default="unknown", min_length=1)
    storage_key: str = Field(default="")
    source_type: str | None = None
    file_size: int | None = None
    metadata_json: str | None = None
    draft_thread_id: str | None = None
    draft_payload: dict[str, Any] | None = None


def get_file_service(db: Session = Depends(get_db)) -> FileService:
    return FileService(db)


@router.post("/{project_id}")
def create_file_record(
    project_id: str,
    payload: FileCreateRequest = Body(...),
    service: FileService = Depends(get_file_service),
) -> dict[str, object]:
    return service.create_file_record(
        project_id=project_id,
        user_id=payload.user_id,
        file_name=payload.file_name,
        file_type=payload.file_type,
        storage_key=payload.storage_key,
        source_type=payload.source_type,
        file_size=payload.file_size,
        metadata_json=payload.metadata_json,
        project_name=payload.project_name,
        draft_thread_id=payload.draft_thread_id,
        draft_payload=payload.draft_payload,
    )


@router.post("/{project_id}/upload")
async def upload_file(
    project_id: str,
    user_id: str = "demo-user-id",
    upload: UploadFile = File(...),
    service: FileService = Depends(get_file_service),
) -> dict[str, object]:
    safe_name = Path(upload.filename or "unnamed").name
    suffix = Path(safe_name).suffix.lower().lstrip(".") or "unknown"
    target_dir = Path(__file__).resolve().parents[4] / "storage" / "projects" / project_id / "files"
    target_dir.mkdir(parents=True, exist_ok=True)
    stored_name = f"{uuid4().hex}_{safe_name}"
    target_path = target_dir / stored_name
    content = await upload.read()
    target_path.write_bytes(content)
    created = service.create_file_record(
        project_id=project_id,
        user_id=user_id,
        file_name=safe_name,
        file_type=suffix,
        storage_key=str(target_path),
        source_type="local_upload",
        file_size=len(content),
        metadata_json=None,
        project_name=Path(safe_name).stem,
    )
    return {**created, "message": "uploaded", "stored_path": str(target_path)}


@router.get("/{project_id}")
def list_files(project_id: str, service: FileService = Depends(get_file_service)) -> dict[str, list[dict[str, object]]]:
    return {"items": service.list_files(project_id)}


@router.get("/{project_id}/{file_id}")
def get_file(project_id: str, file_id: str, service: FileService = Depends(get_file_service)) -> dict[str, object]:
    try:
        return service.get_file(file_id, project_id)
    except ValueError as exc:
        raise not_found(str(exc)) from exc


@router.delete("/{project_id}/{file_id}")
def delete_file(project_id: str, file_id: str, service: FileService = Depends(get_file_service)) -> dict[str, str]:
    try:
        service.delete_file(file_id, project_id)
        return {"message": "deleted"}
    except ValueError as exc:
        raise not_found(str(exc)) from exc
